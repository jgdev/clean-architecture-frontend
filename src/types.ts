export type PaginatedParams = {
  limit?: number;
  skip?: number;
  searchParams?: any;
};

export type PaginatedResult<T> = PaginatedParams & {
  result: T[];
};

export type Operation = {
  id: string;
  type: string;
  name: string;
};

export type Record = {
  id: string;
  operationId: string;
  operationType: string;
  cost: number;
  operationArgs: any[];
  operationResult: any;
  date: Date;
  oldUserBalance: number;
  newUserBalance: number;
};

export type User = {
  email: string;
  balance: number;
};

export type FetchAction<T, K = any> = {
  loading: boolean;
  error?: any;
  result?: T | undefined;
  getAction: (options?: RequestInit & { querystring?: K }) => Promise<any>;
  postAction: (options?: RequestInit & { data?: K }) => Promise<any>;
  delAction: (options?: RequestInit & { data?: K }) => Promise<any>;
};

export type FetchResources<T> = FetchAction<
  PaginatedResult<T>,
  PaginatedParams
>;
