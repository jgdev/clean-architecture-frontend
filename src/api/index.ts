import { User, Record, Operation, ReqParams } from "../types";
import { createFetchAction, createFetchResourcesAction } from "../utils/api";

export const user = createFetchAction<User>("/v1/profile", undefined, {
  response: ({ user }) => user,
});
export const records = createFetchResourcesAction<Record>("/v1/records");
export const operations =
  createFetchResourcesAction<Operation>("/v1/operations");
export const authSession = createFetchAction<
  string,
  { email: string; password: string }
>("/v1/auth/sign-in", window.localStorage.getItem("auth-session-id") || "");

export const deleteRecord = createFetchAction<void>(
  "/v1/records/:recordId",
  undefined,
  {
    url: (url: string, options?: ReqParams<{ recordId: string }>) =>
      url.replace(":recordId", options?.params.recordId),
  }
);

export const authLogout = createFetchAction<void>("/v1/auth/sign-out");
