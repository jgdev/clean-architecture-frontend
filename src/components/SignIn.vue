<script lang="ts" setup>
import { ref, watch } from 'vue'
import { Input, Button } from 'flowbite-vue'
import { FetchAction } from '../types'

type Props = {
  authSession: FetchAction<string>
}

const { authSession } = defineProps<Props>()

const email = ref('')
const password = ref('')

const handleSubmit = (e: Event) => {
  e.preventDefault();

  authSession.error = ''
  authSession.postAction({
    data: {
      email: email.value, password: password.value
    }
  })
}

watch(email, () => {
  authSession.error = ''
})
watch(password, () => {
  authSession.error = ''
})
</script>
<template>
  <div className="sign-in flex flex-col justify-center items-center sm:p-4 sm:max-w-sm m-auto">
    <h1 class="text-xl mb-8">Arithmetic App</h1>
    <div class="rounded-lg p-8 bg-white w-full text-center">
      <h1 class="text-xl mb-8 text-gray-600">Sign in</h1>
      <form @submit="handleSubmit">
        <Input v-model="email" type="email" placeholder="Email" autocomplete="email" required class="mb-4"
          :disabled="authSession.loading" />
        <Input v-model="password" type="password" placeholder="Password" autocomplete="current-password" required
          class="mb-2" :disabled="authSession.loading" />
        <h3 class="my-4 p-0 text-sm text-red-600">{{ authSession.error?.detail || authSession.error?.msg ||
          authSession.error }}</h3>
        <Button color="light" type="submit" :disabled="authSession.loading">{{ authSession.loading ? 'Please wait ...' :
          'Submit'
        }}</Button>
      </form>
    </div>
  </div>
</template>

<style scoped>
.sign-in {
  min-height: 100vh;
}
</style>