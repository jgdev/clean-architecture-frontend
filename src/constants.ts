export const DEFAULT_RESULTS_LIMIT = 10;
