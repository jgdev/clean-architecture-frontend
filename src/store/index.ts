import { ref } from "vue";

export const sessionId = ref(
  window.localStorage.getItem("auth-session-id") || "xd"
);
