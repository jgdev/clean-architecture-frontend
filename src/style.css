@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  font-family: 'SF Pro Display', Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
  line-height: 1.5;
  font-weight: 400;
  font-size: 18px;

  /* color-scheme: light dark;
  background-color: #242424; */
  background-color: white;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

body {
  min-width: 100vw;
  min-height: 100svh;
}

#app {
  min-width: 100vw;
  min-height: 100svh;
  overflow-y: hidden;
}