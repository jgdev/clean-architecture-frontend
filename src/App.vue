<script setup lang="ts">
import SignIn from './components/SignIn.vue';
import ListRecords from './components/ListRecords.vue';
import { onMounted, watch } from 'vue';
import { Record } from './types'
import { operations, records, authSession, user, deleteRecord, authLogout } from './api';
import { DEFAULT_RESULTS_LIMIT } from './constants'

const getRecords = (params: any = {}) => records.getAction({
  querystring: {
    limit: DEFAULT_RESULTS_LIMIT,
    skip: 0,
    ...params,
  }
})

const onDeleteRecord = async (recordId: string, params: any) => {
  await deleteRecord.delAction({
    params: {
      recordId
    }
  })
  await getRecords(params)
}

const onPerformOperation = async (operationType: string, args: any[]) => {
  const operationId = (operations.result?.result || []).find(operation => operation.type === operationType)?.id
  if (!operationId) throw { detail: 'Operation not supported' }
  return records.postAction({
    data: {
      operationId,
      operationArgs: args
    }
  }).then((record: Record) => {
    user.result = { ...user.result!, balance: record.newUserBalance }
    return getRecords()
  })
}

const checkResourceAuthError = (error: any, from?: string) => {
  if (error && error && error.statusCode === 401 || error?.constructor?.name === 'TypeError' && from === 'profile') {
    authSession.result = ''
    window.localStorage.clear()
  }
}

const checkSession = () => {
  if (authSession.result) {
    user.getAction().then(() => {
      getRecords()
      operations.getAction()
    })
  }
}

onMounted(() => {
  checkSession();

  watch(() => authSession.result, () => {
    checkSession();
    localStorage.setItem('auth-session-id', authSession.result || '')
  })
  watch(() => user.error, () => checkResourceAuthError(user.error, 'profile'))
  watch(() => records.error, () => checkResourceAuthError(records.error))
  watch(() => operations.error, () => checkResourceAuthError(operations.error))
})
</script>

<template>
  <SignIn v-if="!authSession.result" :auth-session="authSession" />
  <ListRecords v-if="!!authSession.result" :auth-session="authSession" :auth-logout="authLogout" :user="user"
    :delete-record="deleteRecord" :records="records" :operations="operations" :on-delete-record="onDeleteRecord"
    :on-perform-operation="onPerformOperation" />
</template>