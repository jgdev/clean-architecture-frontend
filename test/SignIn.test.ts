import { mount } from "@vue/test-utils";
import SignIn from "../src/components/SignIn.vue";
import { FetchAction } from "../src/types";

describe("SignIn", () => {
  test("should render properly", () => {
    const authSession = {};

    const wrapper = mount(SignIn, {
      props: {
        authSession,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should disable inputs and buttons on loading", async () => {
    const authSession = {
      loading: true,
    };

    const wrapper = mount(SignIn, {
      props: {
        authSession,
      },
    });

    const submitButton = wrapper.find('button[type="submit"]');
    const inputEmail = wrapper.find('input[type="email"]');
    const inputPassword = wrapper.find('input[type="password"]');

    expect(submitButton.text()).toBe("Please wait ...");

    expect(submitButton.attributes()).toHaveProperty("disabled");
    expect(inputEmail.attributes()).toHaveProperty("disabled");
    expect(inputPassword.attributes()).toHaveProperty("disabled");
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should call login method on submit", async () => {
    const authSession = {
      loading: false,
      postAction: async () => {},
    } as FetchAction<string>;

    const loginSpy = vi.spyOn(authSession, "postAction");

    const wrapper = mount(SignIn, {
      props: {
        authSession,
      },
    });

    const inputEmail = wrapper.find('input[type="email"]');
    const inputPassword = wrapper.find('input[type="password"]');

    await inputEmail.setValue("test@test");
    await inputPassword.setValue("test123");

    await wrapper.find("form").trigger("submit");

    expect(loginSpy).toHaveBeenNthCalledWith(1, {
      data: {
        email: "test@test",
        password: "test123",
      },
    });
  });

  test("should display error properly", async () => {
    const authSession = {
      loading: false,
      error: "Test error",
    } as FetchAction<string>;

    const wrapper = mount(SignIn, {
      props: {
        authSession,
      },
    });

    const error = wrapper.find("h3");

    expect(error.text()).toBe("Test error");

    expect(wrapper.html()).toMatchSnapshot();
  });
});
