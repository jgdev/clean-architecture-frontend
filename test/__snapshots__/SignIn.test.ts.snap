// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignIn > should disable inputs and buttons on loading 1`] = `
"<div data-v-9441a41d=\\"\\" class=\\"sign-in flex flex-col justify-center items-center sm:p-4 sm:max-w-sm m-auto\\">
  <h1 data-v-9441a41d=\\"\\" class=\\"text-xl mb-8\\">Arithmetic App</h1>
  <div data-v-9441a41d=\\"\\" class=\\"rounded-lg p-8 bg-white w-full text-center\\">
    <h1 data-v-9441a41d=\\"\\" class=\\"text-xl mb-8 text-gray-600\\">Sign in</h1>
    <form data-v-9441a41d=\\"\\">
      <div data-v-9441a41d=\\"\\" placeholder=\\"Email\\" autocomplete=\\"email\\" required=\\"\\" class=\\"mb-4\\">
        <!---->
        <div class=\\"flex relative\\">
          <!----><input placeholder=\\"Email\\" autocomplete=\\"email\\" required=\\"\\" class=\\"mb-4 bg-gray-100 border border-gray-300 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 w-full dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 p-2.5 cursor-not-allowed\\" disabled=\\"\\" type=\\"email\\">
          <!---->
        </div>
        <!---->
      </div>
      <div data-v-9441a41d=\\"\\" placeholder=\\"Password\\" autocomplete=\\"current-password\\" required=\\"\\" class=\\"mb-2\\">
        <!---->
        <div class=\\"flex relative\\">
          <!----><input placeholder=\\"Password\\" autocomplete=\\"current-password\\" required=\\"\\" class=\\"mb-2 bg-gray-100 border border-gray-300 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 w-full dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 p-2.5 cursor-not-allowed\\" disabled=\\"\\" type=\\"password\\">
          <!---->
        </div>
        <!---->
      </div>
      <h3 data-v-9441a41d=\\"\\" class=\\"my-4 p-0 text-sm text-red-600\\"></h3><button data-v-9441a41d=\\"\\" class=\\"text-gray-900 bg-white border-gray-300 focus:outline-none focus:ring-4 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 cursor-not-allowed opacity-50 px-4 py-2 text-gray-900 bg-white border border-gray-300 focus:outline-none focus:ring-4 focus:ring-gray-200 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 cursor-not-allowed opacity-50 text-sm px-4 py-2\\" disabled=\\"\\" type=\\"submit\\">
        <!----><span class=\\"\\"><!---->Please wait ...<!----></span>
        <!---->
      </button>
    </form>
  </div>
</div>"
`;

exports[`SignIn > should display error properly 1`] = `
"<div data-v-9441a41d=\\"\\" class=\\"sign-in flex flex-col justify-center items-center sm:p-4 sm:max-w-sm m-auto\\">
  <h1 data-v-9441a41d=\\"\\" class=\\"text-xl mb-8\\">Arithmetic App</h1>
  <div data-v-9441a41d=\\"\\" class=\\"rounded-lg p-8 bg-white w-full text-center\\">
    <h1 data-v-9441a41d=\\"\\" class=\\"text-xl mb-8 text-gray-600\\">Sign in</h1>
    <form data-v-9441a41d=\\"\\">
      <div data-v-9441a41d=\\"\\" placeholder=\\"Email\\" autocomplete=\\"email\\" required=\\"\\" class=\\"mb-4\\">
        <!---->
        <div class=\\"flex relative\\">
          <!----><input placeholder=\\"Email\\" autocomplete=\\"email\\" required=\\"\\" class=\\"mb-4 bg-gray-50 border-gray-300 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 w-full dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 p-2.5\\" type=\\"email\\">
          <!---->
        </div>
        <!---->
      </div>
      <div data-v-9441a41d=\\"\\" placeholder=\\"Password\\" autocomplete=\\"current-password\\" required=\\"\\" class=\\"mb-2\\">
        <!---->
        <div class=\\"flex relative\\">
          <!----><input placeholder=\\"Password\\" autocomplete=\\"current-password\\" required=\\"\\" class=\\"mb-2 bg-gray-50 border-gray-300 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 w-full dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 p-2.5\\" type=\\"password\\">
          <!---->
        </div>
        <!---->
      </div>
      <h3 data-v-9441a41d=\\"\\" class=\\"my-4 p-0 text-sm text-red-600\\">Test error</h3><button data-v-9441a41d=\\"\\" class=\\"text-gray-900 bg-white border-gray-300 focus:outline-none focus:ring-4 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 px-4 py-2 text-gray-900 bg-white border border-gray-300 focus:outline-none focus:ring-4 focus:ring-gray-200 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 text-sm px-4 py-2\\" type=\\"submit\\">
        <!----><span class=\\"\\"><!---->Submit<!----></span>
        <!---->
      </button>
    </form>
  </div>
</div>"
`;

exports[`SignIn > should render properly 1`] = `
"<div data-v-9441a41d=\\"\\" class=\\"sign-in flex flex-col justify-center items-center sm:p-4 sm:max-w-sm m-auto\\">
  <h1 data-v-9441a41d=\\"\\" class=\\"text-xl mb-8\\">Arithmetic App</h1>
  <div data-v-9441a41d=\\"\\" class=\\"rounded-lg p-8 bg-white w-full text-center\\">
    <h1 data-v-9441a41d=\\"\\" class=\\"text-xl mb-8 text-gray-600\\">Sign in</h1>
    <form data-v-9441a41d=\\"\\">
      <div data-v-9441a41d=\\"\\" placeholder=\\"Email\\" autocomplete=\\"email\\" required=\\"\\" class=\\"mb-4\\">
        <!---->
        <div class=\\"flex relative\\">
          <!----><input placeholder=\\"Email\\" autocomplete=\\"email\\" required=\\"\\" class=\\"mb-4 bg-gray-50 border-gray-300 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 w-full dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 p-2.5\\" type=\\"email\\">
          <!---->
        </div>
        <!---->
      </div>
      <div data-v-9441a41d=\\"\\" placeholder=\\"Password\\" autocomplete=\\"current-password\\" required=\\"\\" class=\\"mb-2\\">
        <!---->
        <div class=\\"flex relative\\">
          <!----><input placeholder=\\"Password\\" autocomplete=\\"current-password\\" required=\\"\\" class=\\"mb-2 bg-gray-50 border-gray-300 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 w-full dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500 p-2.5\\" type=\\"password\\">
          <!---->
        </div>
        <!---->
      </div>
      <h3 data-v-9441a41d=\\"\\" class=\\"my-4 p-0 text-sm text-red-600\\"></h3><button data-v-9441a41d=\\"\\" class=\\"text-gray-900 bg-white border-gray-300 focus:outline-none focus:ring-4 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 px-4 py-2 text-gray-900 bg-white border border-gray-300 focus:outline-none focus:ring-4 focus:ring-gray-200 font-medium rounded-lg dark:bg-gray-800 dark:text-white dark:border-gray-600 dark:focus:ring-gray-700 hover:bg-gray-100 dark:hover:border-gray-600 text-sm px-4 py-2\\" type=\\"submit\\">
        <!----><span class=\\"\\"><!---->Submit<!----></span>
        <!---->
      </button>
    </form>
  </div>
</div>"
`;
